sample-apps/go-sample-app/collection/client.go

```diff
@@ -0,0 +1,148 @@
+package collection
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/contrib/propagators/aws/xray"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/metric/global"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"google.golang.org/grpc"
+
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/histogram"
+	controller "go.opentelemetry.io/otel/sdk/metric/controller/basic"
+	processor "go.opentelemetry.io/otel/sdk/metric/processor/basic"
+	selector "go.opentelemetry.io/otel/sdk/metric/selector/simple"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+var cfg = GetConfiguration()
+
+const grpcEndpoint = "0.0.0.0:4317"
+
+const serviceName = "go"
+
+var tracer = otel.Tracer("github.com/aws-otel-commnunity/sample-apps/go-sample-app/collection")
+
+// Names for metric instruments
+const apiTimeAlive = "timeAlive"
+const apiCpuUsage = "cpuUsage"
+const apiTotalHeapSize = "totalHeapSize"
+const apiThreadsActive = "threadsActive"
+const apiTotalBytesSent = "totalBytesSent"
+const apiTotalApiRequests = "totalApiRequests"
+const apiLatencyTime = "latencyTime"
+
+// Common attributes for traces and metrics (random, request)
+var requestMetricCommonLabels = []attribute.KeyValue{
+	attribute.String("signal", "metric"),
+	attribute.String("language", serviceName),
+	attribute.String("metricType", "request"),
+}
+
+var randomMetricCommonLabels = []attribute.KeyValue{
+	attribute.String("signal", "metric"),
+	attribute.String("language", serviceName),
+	attribute.String("metricType", "random"),
+}
+
+var traceCommonLabels = []attribute.KeyValue{
+	attribute.String("signal", "trace"),
+	attribute.String("language", serviceName),
+	attribute.String("host", cfg.Host),
+	attribute.String("port", cfg.Port),
+}
+
+// StartClient starts the OTEL controller which periodically collects signals and exports them.
+// Trace exporter and Metric exporter are both configured.
+func StartClient(ctx context.Context) (func(context.Context) error, error) {
+
+	// Setup trace related
+	tp, err := setupTraceProvider(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(xray.Propagator{}) // Set AWS X-Ray propagator
+
+	// Setup metric related
+	ctrl, err := setupMetricsController(ctx)
+	if err != nil {
+		return nil, err
+	}
+	global.SetMeterProvider(ctrl)
+
+	return func(context.Context) error {
+		ctx, cancel := context.WithTimeout(ctx, time.Second)
+		defer cancel()
+
+		defer func() error {
+			tpErr := tp.Shutdown(ctx)
+			if tpErr != nil {
+				return tpErr
+			}
+			return nil
+		}()
+		// pushes any last exports to the receiver
+		return ctrl.Stop(ctx)
+	}, nil
```

Comment:
```suggestion
	return func(context.Context) (err error) {
		ctx, cancel := context.WithTimeout(ctx, time.Second)
		defer cancel()

		defer func() {
			tpErr := tp.Shutdown(ctx)
			if tpErr != nil {
				err = tpErr
			}
		}()

		// pushes any last exports to the receiver
		err = ctrl.Stop(ctx)
		return
	}, nil
```

Dealing with errors from deferred functions is a bit weird.  The function where the `defer` invocation happens needs to have a named error return value if you want to be able to modify it in a deferred invocation.  Then you don't return an error from the deferred invocation, but change the captured variable that will be returned from the outer function.

Reply:
I agree. The way deferred functions work kind of confuses me but this makes a lot of sense. I updated the code